### PR TITLE
fix avatar-mixer crash during sort

### PR DESCRIPTION
**Pull request by AndrewMeadows**

### CODE REVIEW

File: assignment-client/src/avatars/AvatarMixerSlave.cpp [hunk 1]
```diff
@@ -33,11 +33,6 @@
 #include "AvatarMixerClientData.h"
 #include "AvatarMixerSlave.h"
 
-namespace PrioritySortUtil {
-    // we declare this callback here but override it later
-    std::function<uint64_t(const AvatarSharedPointer&)> getAvatarAgeCallback = [] (const AvatarSharedPointer& avatar) { return 0; };
-}
```
Comment (AndrewMeadows) [inline assignment-client/src/avatars/AvatarMixerSlave.cpp]: @Atlante45 figured out the problem: PrioritySortUtil::getAvatarAgeCallback is a global so multithreaded avatar-mixers would overwrite this on top of each other.

File: assignment-client/src/avatars/AvatarMixerSlave.cpp [hunk 2]
```diff
@@ -200,34 +196,29 @@ void AvatarMixerSlave::broadcastAvatarDataToAgent(const SharedNodePointer& node)
             AvatarSharedPointer otherAvatar = otherNodeData->getAvatarSharedPointer();
             avatarsToSort.push_back(otherAvatar);
             avatarDataToNodes[otherAvatar] = otherNode;
+            QUuid id = otherAvatar->getSessionUUID();
+            avatarEncodeTimes[id] = nodeData->getLastOtherAvatarEncodeTime(id);
         }
     });
```
Comment (AndrewMeadows) [inline assignment-client/src/avatars/AvatarMixerSlave.cpp]: We now also build a map of encode times.

File: assignment-client/src/avatars/AvatarMixerSlave.cpp [hunk 3]
```diff
@@ -200,34 +196,29 @@ void AvatarMixerSlave::broadcastAvatarDataToAgent(const SharedNodePointer& node)
...
-            // use the callback implemented above
-            return PrioritySortUtil::getAvatarAgeCallback(_avatar);
+            return _lastEncodeTime;
         }
         const AvatarSharedPointer& getAvatar() const { return _avatar; }
 
     private:
         AvatarSharedPointer _avatar;
+        uint64_t _lastEncodeTime;
     };
```
Comment (AndrewMeadows) [inline assignment-client/src/avatars/AvatarMixerSlave.cpp]: `SortableAvatar` takes a `lastEncodeTime` in its ctor.

File: assignment-client/src/avatars/AvatarMixerSlave.cpp [hunk 4]
```diff
@@ -322,7 +313,12 @@ void AvatarMixerSlave::broadcastAvatarDataToAgent(const SharedNodePointer& node)
 
         if (!shouldIgnore) {
             // sort this one for later
-            sortedAvatars.push(SortableAvatar(avatar));
+            uint64_t lastEncodeTime = 0;
+            std::unordered_map<QUuid, uint64_t>::const_iterator itr = avatarEncodeTimes.find(avatar->getSessionUUID());
+            if (itr != avatarEncodeTimes.end()) {
+                lastEncodeTime = itr->second;
+            }
+            sortedAvatars.push(SortableAvatar(avatar, lastEncodeTime));
```
Comment (AndrewMeadows) [inline assignment-client/src/avatars/AvatarMixerSlave.cpp]: We look up the correct `lastEncodeTime` whenever we create a new `SortableAvatar`.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5581/HighFidelity-Beta-PR11962-db6505493f212347dc93891dee060aaf00474732.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5581/HighFidelity-Beta-PR11962-db6505493f212347dc93891dee060aaf00474732.dmg

**Comment by ZappoMan** [comment]
Test Plan:
- Run this SERVER build on a multi-threaded server...
- Connect lots of clients
- Wait around and hope it doesn't crash.
- If it crashes it's a fail.

**Comment by RyanDowne** [comment]
You can use dev-welcome for this test.

**Comment by AndrewMeadows** [comment]
@sethalves performed the multi-threaded test with no crashes.  Perhaps he can add the **QA_APPROVED** label.

I checked this avatar-mixer to make sure sorting/throttling still works (in a single-threaded environment) and found it working correctly.

**Comment by AndrewMeadows** [comment]
Does QA have the capabilities to deploy to dev-welcome?  Or do they ask someone to please do it?  Who should be taking initiative on this?

**Pull request merged**